File: app/src/main/java/com/prasunmondal/libs/Logs/LogUtils.java
package com.prasunmondal.libs.Logs;

public class LogUtils {

    public static StackTraceElement getCaller() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        return stackTraceElements[getCallerIndex(stackTraceElements)];
    }

    static int getCallerIndex(StackTraceElement[] stackTraceElements) {
        int i;
        boolean enteredLogMe = false;
        for (i = 0; i < stackTraceElements.length; i++) {
            if (stackTraceElements[i].getClassName().startsWith(LogMe.class.getPackage().getName())) {
                enteredLogMe = true;
            }
            if (enteredLogMe && !stackTraceElements[i].getClassName().startsWith(LogMe.class.getPackage().getName())) {
                break;
            }
        }
        return i;
    }

    static String getPrefix() {
        StackTraceElement caller = getCaller();
        return LogConfigurations.APP_LOGS_PREFIX + getClassName(caller) + ":" + getLineNumber(caller) + "|" + getMethodName(caller) + "(): ";
    }

    static String putPrefix(String str) {
        String prefix = getPrefix();
        str = str.replace("\n", "\n" + prefix);
        return str;
    }

    static String putOffset(String str) {
        String lineOffset = lineOffset();
        str = str.replace("\n", "\n" + lineOffset);
        str = lineOffset + str;
        return str;
    }

    static String lineOffset() {
        if (!LogConfigurations.IS_LINE_OFFSET_ENABLED) {
            return "";
        }

        StringBuilder str = new StringBuilder();
        int logDepth = getdepth();
        for (int i = 0; i < logDepth; i++) {
            str.append(LogConfigurations.OFFSET_STRING_UNIT);
        }
        return str.toString();
    }

    private static String getClassName(StackTraceElement caller) {
        String className = caller.getClassName();
        return className.substring(className.lastIndexOf('.') + 1);
    }

    private static int getLineNumber(StackTraceElement caller) {
        return caller.getLineNumber();
    }

    private static String getMethodName(StackTraceElement caller) {
        String methodName = caller.getMethodName();
        if (methodName.equals("<init>")) {
            methodName = LogConfigurations.CONSTRUCTOR_INDICATOR;
        }
        return methodName;
    }

    static int getdepth() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        int i = getCallerIndex(stackTraceElements);
        int j = 0;
        for (; i < stackTraceElements.length; i++) {
            if (stackTraceElements[i].toString().startsWith("android.")) {
                return j;
            } else {
                j++;
            }
        }
        return j;
    }
}

File: app/src/main/java/com/prasunmondal/libs/logs/LogUtils.kt
package com.prasunmondal.libs.logs

import com.prasunmondal.libs.logs.instant.terminal.LogMe
import com.prasunmondal.libs.reflections.code.stacktrace.GetStacktraceDetails

object LogUtils {
    val caller: StackTraceElement
        get() {
            val stackTraceElements = Thread.currentThread().stackTrace
            return stackTraceElements[getCallerIndex(stackTraceElements)]
        }

    @JvmStatic
    fun getCallerIndex(stackTraceElements: Array<StackTraceElement>): Int {
        var i: Int
        var enteredLogMe = false
        i = 0
        while (i < stackTraceElements.size) {
            if (stackTraceElements[i].className.startsWith(LogMe::class.java.getPackage().name)) {
                enteredLogMe = true
            }
            if (enteredLogMe && !stackTraceElements[i].className.startsWith(LogMe::class.java.getPackage().name)) {
                break
            }
            i++
        }
        return i
    }

    @JvmStatic
    val prefix: String
        get() {
            val caller = caller
            return LogConfigurations.APP_LOGS_PREFIX + GetStacktraceDetails.className(caller) + ":" + GetStacktraceDetails.getLineNumber(
                caller
            ) + "|" + GetStacktraceDetails.getMethodName(caller) + "(): "
        }

    @JvmStatic
    fun putPrefix(str: String): String {
        var str = str
        val prefix = prefix
        str = str.replace(
            "\n", """
     
     $prefix
     """.trimIndent()
        )
        return str
    }

    fun putOffset(str: String): String {
        var str = str
        val lineOffset = lineOffset()
        str = str.replace(
            "\n", """
     
     $lineOffset
     """.trimIndent()
        )
        str = lineOffset + str
        return str
    }

    fun lineOffset(): String {
        if (!LogConfigurations.IS_LINE_OFFSET_ENABLED) {
            return ""
        }
        val str = StringBuilder()
        val logDepth = getdepth()
        for (i in 0 until logDepth) {
            str.append(LogConfigurations.OFFSET_STRING_UNIT)
        }
        return str.toString()
    }

    fun getdepth(): Int {
        val stackTraceElements = Thread.currentThread().stackTrace
        var i = getCallerIndex(stackTraceElements)
        var j = 0
        while (i < stackTraceElements.size) {
            if (stackTraceElements[i].toString().startsWith("android.")) {
                return j
            } else {
                j++
            }
            i++
        }
        return j
    }
}
